File: docs/chapter1/section4/src/practice_advanced.go
package main

import (
	"database/sql"
	"errors"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/go-sql-driver/mysql"
	"github.com/jmoiron/sqlx"
)

type City struct {
	ID          int    `json:"ID,omitempty" db:"ID"`
	Name        string `json:"name,omitempty" db:"Name"`
	CountryCode string `json:"countryCode,omitempty"  db:"CountryCode"`
	District    string `json:"district,omitempty"  db:"District"`
	Population  int    `json:"population,omitempty"  db:"Population"`
}

func main() {
	jst, err := time.LoadLocation("Asia/Tokyo")
	if err != nil {
		log.Fatal(err)
	}

	conf := mysql.Config{
		User:      os.Getenv("DB_USERNAME"),
		Passwd:    os.Getenv("DB_PASSWORD"),
		Net:       "tcp",
		Addr:      os.Getenv("DB_HOSTNAME") + ":" + os.Getenv("DB_PORT"),
		DBName:    os.Getenv("DB_DATABASE"),
		ParseTime: true,
		Collation: "utf8mb4_unicode_ci",
		Loc:       jst,
	}

	db, err := sqlx.Open("mysql", conf.FormatDSN())

	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("connected")

	cityName := os.Args[1]

	var city City
	err = db.Get(&city, "SELECT * FROM city WHERE Name = ?", cityName)
	if errors.Is(err, sql.ErrNoRows) {
		log.Printf("no such city Name = '%s'\n", cityName)
		return
	} else if err != nil {
		log.Fatalf("DB Error: %s\n", err)
	}

	fmt.Printf("%sの人口は%d人です\n", city.Name, city.Population)

	var population int                                                                           //[!code ++]
	err = db.Get(&population, "SELECT Population FROM country WHERE Code = ?", city.CountryCode) //[!code ++]
	if errors.Is(err, sql.ErrNoRows) {                                                           //[!code ++]
		log.Printf("no such country Code = '%s'\n", city.CountryCode) //[!code ++]
		return                                                        //[!code ++]
	} else if err != nil { //[!code ++]
		log.Fatalf("DB Error: %s\n", err) //[!code ++]
	} //[!code ++]
	//[!code ++]
	percent := (float64(city.Population) / float64(population)) * 100 //[!code ++]
	//[!code ++]
	fmt.Printf("これは%sの人口の%f%%です\n", city.CountryCode, percent) //[!code ++]
}

File: docs/chapter1/section4/src/practice_advanced.rs
use anyhow::Ok;
use sqlx::mysql::MySqlConnectOptions;
use std::env;

#[derive(sqlx::FromRow)]
#[sqlx(rename_all = "PascalCase")]
#[allow(dead_code)] // 使用していないフィールドへの警告を抑制
struct City {
    #[sqlx(rename = "ID")]
    id: i32,
    name: String,
    country_code: String,
    district: String,
    population: i32,
}

fn get_option() -> anyhow::Result<MySqlConnectOptions> {
    let host = env::var("DB_HOSTNAME")?;
    let port = env::var("DB_PORT")?.parse()?;
    let username = env::var("DB_USERNAME")?;
    let password = env::var("DB_PASSWORD")?;
    let database = env::var("DB_DATABASE")?;
    let timezone = Some(String::from("Asia/Tokyo"));
    let collation = String::from("utf8mb4_unicode_ci");

    Ok(MySqlConnectOptions::new()
        .host(&host)
        .port(port)
        .username(&username)
        .password(&password)
        .database(&database)
        .timezone(timezone)
        .collation(&collation))
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let options = get_option()?;
    let pool = sqlx::MySqlPool::connect_with(options).await?;

    let city_name = env::args().nth(1).expect("city name is required");
    println!("Connected");
    let city = sqlx::query_as::<_, City>("SELECT * FROM city WHERE Name = ?")
        .bind(&city_name)
        .fetch_one(&pool)
        .await
        .map_err(|e| match e {
            sqlx::Error::RowNotFound => anyhow::anyhow!("no such city Name = {}\n", &city_name),
            _ => anyhow::anyhow!("DB error: {}", e),
        })?;

    println!("{}の人口は{}人です", &city.name, &city.population);

    let population: i64 = sqlx::query_scalar("SELECT Population FROM country WHERE Code = ?") // [!code ++]
        .bind(&city.country_code)  // [!code ++]
        .fetch_one(&pool)  // [!code ++]
        .await  // [!code ++]
        .map_err(|e| match e {  // [!code ++]
            sqlx::Error::RowNotFound => { // [!code ++]
                anyhow::anyhow!("no such country Code = {}\n", &city.country_code) // [!code ++]
            } // [!code ++]
            _ => anyhow::anyhow!("DB error: {}", e), // [!code ++]
        })?; // [!code ++]
    let percent = city.population as f64 / population as f64 * 100.0; // [!code ++]
    println!("これは、{}の人口の{:.2}%です", &city.country_code, percent); // [!code ++]

    Ok(())
}
